pydantic_settings/sources.py

```diff
@@ -1912,7 +1912,15 @@ def _read_files(self, files: PathType | None) -> dict[str, Any]:
         for file in files:
             file_path = Path(file).expanduser()
             if file_path.is_file():
-                vars.update(self._read_file(file_path))
+                try:
+                    settings = self._read_file(file_path)
+                except ValueError as e:
+                    raise SettingsError(f'Failed to parse settings from {file_path}, {e}')
+                if not isinstance(settings, dict):
+                    raise SettingsError(
+                        f'Failed to parse settings from {file_path}, expecting an object (valid dictionnary)'
+                    )
+                vars.update(settings)
```

Comment:
@l00ptr Thanks for updating this.
As I mentioned before we can't merge this PR because it will introduce a breaking change. So, we have 2 options here:

1- Keep only the tests that test the current behavior(ValueError) and revert to raising `SettingsError`; then, we can merge the tests only.
2- Keep the PR open until `V3` and merge it before `V3`.

I would prefer option 1.

Reply:
@hramezani i splitted those changes on 3  commits, let me know if i should create a specific MR for the 2 last commits (about SettingsError) and only keep the first one here.

Reply:
@l00ptr no need for a new PR.

Then we have to wait to merge it on `V3`. I will add a `v3` label to the PR to not merge it until `v3`